glean-core/ios/Glean/Dispatchers.swift

```diff
@@ -88,17 +83,62 @@ class Dispatchers {
+            serialOperationQueue.addOperation(block)
+
+            // If we are in testing mode, go ahead and wait until it is finished before continuing
+            // to ensure synchronous execution of the task.
             if testingMode {
-                // If we are in testing mode, go ahead and execute the operation and wait until it
-                // is finished before continuing to give us synchronous execution of the task.
-                operation.start()
-                operation.waitUntilFinished()
-            } else {
-                operationQueue.addOperation(operation)
+                serialOperationQueue.waitUntilAllOperationsAreFinished()
             }
         }
+    }
+
+    /// This function launches a background `Operation` on a concurrently executed queue.
+    ///
+    /// - parameters:
+    ///   * operation: The `Operation` to execute
+    ///
+    /// This function is used to execute tasks in an asynchrounous manner and still give us the ability
+    /// to cancel the tasks by creating them as `Operation`s rather than using GCD.
+    ///
+    /// This function specifically ignores the `queueInitialTasks` flag because the only tasks that
+    /// should be launched by this are the ping upload schedulers and those should run regardless of
+    /// the initialized state.
+    ///
+    /// If `testingMode` is enabled, then launch will just execute the task rather than adding it to the
+    /// concurrent queue to avoid asynchronous issues while testing
+    func launchConcurrent(operation: Operation) {
+        if testingMode {
+            operation.start()
+            operation.waitUntilFinished()
+        } else {
+            concurrentOperationsQueue.addOperation(operation)
+        }
+    }
+
+    /// This function launches a background block of code on a concurrently executed queue.
+    ///
+    /// - parameters:
+    ///   * block: The block of code to execute, as a Closure that accepts no arugments and returns Void
+    ///
+    /// This function is used to execute tasks in an asynchrounous manner and still give us the ability
+    /// to cancel the tasks by creating them as `Operation`s rather than using GCD.
+    ///
+    /// This function specifically ignores the `queueInitialTasks` flag because the only tasks that
+    /// should be launched by this are the ping upload schedulers and those should run regardless of
+    /// the initialized state.
+    ///
+    /// If `testingMode` is enabled, then launch will just execute the task rather than adding it to the
+    /// concurrent queue to avoid asynchronous issues while testing
+    func launchConcurrent(block: @escaping () -> Void) {
+        launchConcurrent(operation: BlockOperation(block: block))
+    }
 
-        return operation
+    /// Cancel any pending background tasks
+    func cancelBackgroundTasks() {
+        serialOperationQueue.cancelAllOperations()
+        concurrentOperationsQueue.cancelAllOperations()
```

Comment:
Note: this will also cancel `sendPingsByName`. Could this result in partial pings or corrupted data?

Reply:
Since this cancels the network request, I would expect this to not perform a partial upload.  Looking at the [docs](https://developer.apple.com/documentation/foundation/urlsessiontask/1411591-cancel) it looks like it notifies the task so that it can cancel gracefully.

Reply:
This is doing ping persistence to disk in addition to the upload. If we cancel (kill the request?) while writing to disk then we result in partial pings being written. Upload doesn't concern me too much

Reply:
Ahh I see what you are asking.  The answer is, I'm pretty sure it's safe, and here is why:  The way that the `sendPingsByName` function launches on the concurrent queue is by passing a closure block to be executed.  Since this block doesn't do anything to handle the cancel state, it would run to completion.  Cancel in this case does not mean that it is aborting the thread, it means that it signals the operation to cancel.  See the test I added and how I used the isCancelled in the small counter operations I used.  If I didn't do that, then they would run forever (As I found out the first time I tried the test without it).

Reply:
Want me try to write a test around this?

Reply:
Ah, I see now. That makes sense, thanks for explaining. Can you add a small comment above the `concurrentOperationsQueue.cancelAllOperations()` line to explain why this is safe (basically, what you just told me). Writing a test for this would be a bit tricky, I'd say not a blocker for now, but file a bug for it if you feel like :)

Reply:
I'll add the comment, and I filed a bug to follow up on making the operation that is calling the function that serializes pings obey the cancel state gracefully [here](https://bugzilla.mozilla.org/show_bug.cgi?id=1582464).